### PR TITLE
Expand JIT groupby test suite

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR reorganizes and expands the test suite for groupby apply functions using the JIT engine to include nan cases and cases where the groups are larger than a single thread block.

### CODE REVIEW

**Comment by brandon-b-miller** [comment]
Marking this ready for review - I am going to work on solving some of the `xfails` that came up when putting this PR, so hopefully some of them won't even be necessary by merge time.

File: python/cudf/cudf/tests/test_groupby.py [hunk 1]
```diff
@@ -429,48 +478,92 @@ def func(df):
...
+    data["val1"] = data["val1"].astype(dtype)
+    data["val2"] = data["val2"].astype(dtype)
+    data["val1"] = special_val
+    data["val2"] = special_val
 
-    run_groupby_apply_jit_test(groupby_jit_data, func, ["key1"])
+    run_groupby_apply_jit_test(data, func, ["key1"])
+
+
+@pytest.mark.parametrize("dtype", ["float64"])
```
Comment (bdice) [inline python/cudf/cudf/tests/test_groupby.py:556]: Should this test float32, too?

**Review by bdice — COMMENTED** [review]
@brandon-b-miller Just to clarify, you're planning to merge this to introduce new xfailed tests, then fix them in subsequent PRs?

**Comment by brandon-b-miller** [comment]
> @brandon-b-miller Just to clarify, you're planning to merge this to introduce new xfailed tests, then fix them in subsequent PRs?

Correct, this is my plan, based on not knowing how long it might take to fix some of these.

File: python/cudf/cudf/tests/test_groupby.py [hunk 2]
```diff
@@ -381,19 +381,76 @@ def emulate(df):
...
+    """
+    Small dataset for testing groupby apply with jit
+    returns a dataframe whose keys columns define
+    4 groups of size 1, 2, 3, 4 as well as an additional
+    key column that can be used to test subgroups
+
+    useful for very basic testing of result values
+
+    """
     np.random.seed(0)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_groupby.py:394]: I realize it's not quite in this PR, but having the seed be set as part of a fixture seems like a confusing and potentially misleading side effect. If you want the fixture to generate a consistent output, can you use a seeded generator instead? Here are the docs on the canonical way to do this: https://numpy.org/doc/stable/reference/random/generator.html

File: python/cudf/cudf/tests/test_groupby.py [hunk 3]
```diff
@@ -381,19 +381,76 @@ def emulate(df):
...
-    df["val1"] = np.random.random(nelem)
-    df["val2"] = np.random.random(nelem)
-    df["val3"] = np.random.randint(0, 10, nelem)
-    df["val4"] = np.random.randint(0, 10, nelem)
+    key1 = (
+        [1]  # group of size 1
+        + [2, 2]  # group of size 2
+        + [3, 3, 3]  # group of size 3
+        + [4, 4, 4, 4]  # group of size 4
+    )
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_groupby.py:401]: Writing it this way should make the group sizes more obvious.
```suggestion
    key1 = (
        [1]
        + [2] * 2
        + [3] * 3
        + [4] * 4
    )
```

File: python/cudf/cudf/tests/test_groupby.py [hunk 4]
```diff
@@ -381,19 +381,76 @@ def emulate(df):
...
+    key2 = [1, 2] * 5
+    df["key1"] = key1
+    df["key2"] = key2
+
+    df["val1"] = np.random.randint(0, 10, len(key1))
+    df["val2"] = np.random.randint(0, 10, len(key1))
+
+    # randomly permute data
+    df = df.sample(frac=1).reset_index(drop=True)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_groupby.py:410]:
```suggestion
    df = df.sample(frac=1, ignore_index=True)
```

File: python/cudf/cudf/tests/test_groupby.py [hunk 5]
```diff
@@ -381,19 +381,76 @@ def emulate(df):
...
+
+@pytest.fixture(scope="module")
+def groupby_jit_data_nans(groupby_jit_data_small):
+    """
+    Small dataset containing nans
+    """
+
+    df = groupby_jit_data_small.sort_values(["key1", "key2"])
+    df["val1"][::2] = np.nan
+    df = df.sample(frac=1).reset_index(drop=True)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_groupby.py:439]:
```suggestion
    df = df.sample(frac=1, ignore_index=True)
```

File: python/cudf/cudf/tests/test_groupby.py [hunk 6]
```diff
@@ -429,48 +478,92 @@ def func(df):
...
+@pytest.mark.parametrize("dataset", ["small", "large", "nans"])
+def test_groupby_apply_jit_unary_reductions(
+    func, dtype, dataset, groupby_jit_datasets
+):
+    dataset = groupby_jit_datasets[dataset]
 
-    groupby_jit_data["val3"] = groupby_jit_data["val3"].astype(dtype)
-    groupby_jit_data["val4"] = groupby_jit_data["val4"].astype(dtype)
+    if func in {"sum", "mean", "var", "std"} and dtype == "int32":
+        with pytest.xfail(
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_groupby.py:503]: Don't use pytest.xfail directly, use pytest.mark.xfail. More explanation: https://docs.rapids.ai/api/cudf/stable/developer_guide/testing/#conditional-xfails

**Review by vyasr — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/tests/test_groupby.py [hunk 7]
```diff
@@ -392,19 +392,72 @@ def emulate(df):
 
 
 @pytest.fixture(scope="module")
-def groupby_jit_data():
-    np.random.seed(0)
+def groupby_jit_data_small():
+    """
+    Small dataset for testing groupby apply with jit
```
Comment (bdice) [inline python/cudf/cudf/tests/test_groupby.py:397]: Can we write this comment with complete sentences / capitalization? It also feels like it's wrapped pretty harshly, like 50-60 chars / line? Same request for other comments below.

File: python/cudf/cudf/tests/test_groupby.py [hunk 8]
```diff
@@ -440,36 +523,151 @@ def func(df):
...
+    dataset["val1"] = dataset["val1"].astype(dtype)
+    dataset["val2"] = dataset["val2"].astype(dtype)
+
+    keys = ["key1"]
+
+    def func(group):
+        return group["val1"].corr(group["val2"])
+
+    if np.dtype(dtype).kind == "f":
+        # Feature Request:
```
Comment (bdice) [inline python/cudf/cudf/tests/test_groupby.py:642]:
```suggestion
        # Correlation of floating types is not yet supported:
```

**Review by bdice — APPROVED** [review]
A couple small cleanup requests, otherwise LGTM!

**Review by vyasr — APPROVED** [review]

**Pull request merged**